src/bijectors/composed.jl

```diff
@@ -0,0 +1,190 @@
+```julia
+cb1 = b1 ∘ b2                  # => Composed.ts == (b2, b1)
+cb2 = composel(b2, b1)         # => Composed.ts == (b2, b1)
+cb1(x) == cb2(x) == b1(b2(x))  # => true
+```
+
+## Structure
+`∘` will result in "flatten" the composition structure while `composel` and
+`composer` preserve the compositional structure. This is most easily seen by an example:
+```julia-repl
+julia> b = Exp()
+Exp{0}()
+
+julia> cb1 = b ∘ b; cb2 = b ∘ b;
+
+julia> (cb1 ∘ cb2).ts # <= different
+(Exp{0}(), Exp{0}(), Exp{0}(), Exp{0}())
+
+julia> (cb1 ∘ cb2).ts isa NTuple{4, Exp{0}}
+true
+
+julia> Bijectors.composer(cb1, cb2).ts
+(Composed{Tuple{Exp{0},Exp{0}},0}((Exp{0}(), Exp{0}())), Composed{Tuple{Exp{0},Exp{0}},0}((Exp{0}(), Exp{0}())))
+
+julia> Bijectors.composer(cb1, cb2).ts isa Tuple{Composed, Composed}
+true
+```
+
+"""
+struct Composed{A, N} <: Bijector{N}
+    ts::A
+
+    Composed(bs::C) where {N, C<:Tuple{Vararg{<:Bijector{N}}}} = new{C, N}(bs)
+    Composed(bs::A) where {N, A<:AbstractArray{<:Bijector{N}}} = new{A, N}(bs)
+end
+
+isclosedform(b::Composed) = all(isclosedform.(b.ts))
+
+
+"""
+    composel(ts::Bijector...)::Composed{<:Tuple}
+
+Constructs `Composed` such that `ts` are applied left-to-right.
+"""
+composel(ts::Bijector{N}...) where {N} = Composed(ts)
+
+"""
+    composer(ts::Bijector...)::Composed{<:Tuple}
+
+Constructs `Composed` such that `ts` are applied right-to-left.
+"""
+composer(ts::Bijector{N}...) where {N} = Composed(reverse(ts))
+
+# The transformation of `Composed` applies functions left-to-right
+# but in mathematics we usually go from right-to-left; this reversal ensures that
+# when we use the mathematical composition ∘ we get the expected behavior.
+# TODO: change behavior of `transform` of `Composed`?
+@generated function ∘(b1::Bijector{N1}, b2::Bijector{N2}) where {N1, N2}
+    if N1 == N2
+        return :(composel(b2, b1))
+    else
+        return :(throw(DimensionMismatch("$(typeof(b1)) expects $(N1)-dim but $(typeof(b2)) expects $(N2)-dim")))
+    end
+end
+
+∘(b1::Composed, b2::Bijector) = composel(b2, b1.ts...)
+∘(b1::Bijector, b2::Composed) = composel(b2.ts..., b1)
+∘(b1::Composed, b2::Composed) = composel(b2.ts..., b1.ts...)
+
+∘(::Identity{N}, ::Identity{N}) where {N} = Identity{N}()
+∘(::Identity{N}, b::Bijector{N}) where {N} = b
+∘(b::Bijector{N}, ::Identity{N}) where {N} = b
+
+inv(ct::Composed) = composer(map(inv, ct.ts)...)
+@generated function inv(cb::Composed{A}) where {A<:Tuple}
+    exprs = []
+
+    # inversion → reversing order
+    for i = length(A.parameters):-1:1
+        push!(exprs, :(inv(cb.ts[$i])))
+    end
+    return :(Composed(($(exprs...), )))
+end
+
+# # TODO: should arrays also be using recursive implementation instead?
+function (cb::Composed{<:AbstractArray{<:Bijector}})(x)
+    res = x
+    for b ∈ cb.ts
+        res = b(res)
+    end
+
+    return res
+end
+
+# recursive implementation like this allows type-inference
+_transform(x, b1::Bijector, b2::Bijector) = b2(b1(x))
```

Comment:
Implement this without recursion using `@generated` like we've done for `forward`.